pkg/storegateway/limiter.go

```diff
@@ -0,0 +1,150 @@
+package storegateway
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/store"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/cortexproject/cortex/pkg/storage/tsdb"
+	"github.com/cortexproject/cortex/pkg/util"
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+type limiter struct {
+	limiter *store.Limiter
+}
+
+func (c *limiter) Reserve(num uint64) error {
+	return c.ReserveWithType(num, 0)
+}
+
+func (c *limiter) ReserveWithType(num uint64, _ store.StoreDataType) error {
+	err := c.limiter.Reserve(num)
+	if err != nil {
+		return httpgrpc.Errorf(http.StatusUnprocessableEntity, err.Error())
+	}
+
+	return nil
+}
+
+type compositeLimiter struct {
+	limiters []store.BytesLimiter
+}
+
+func (c *compositeLimiter) ReserveWithType(num uint64, dataType store.StoreDataType) error {
+	for _, l := range c.limiters {
+		if err := l.ReserveWithType(num, dataType); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type tokenBucketLimiter struct {
+	instanceTokenBucket *util.TokenBucket
+	userTokenBucket     *util.TokenBucket
+	requestTokenBucket  *util.TokenBucket
+	getTokensToRetrieve func(tokens uint64, dataType store.StoreDataType) int64
+	dryRun              bool
+}
+
+func (t *tokenBucketLimiter) Reserve(_ uint64) error {
+	return nil
+}
+
+func (t *tokenBucketLimiter) ReserveWithType(num uint64, dataType store.StoreDataType) error {
+	tokensToRetrieve := t.getTokensToRetrieve(num, dataType)
+
+	// check request bucket
+	retrieved := t.requestTokenBucket.Retrieve(tokensToRetrieve)
+	if retrieved {
+		t.userTokenBucket.ForceRetrieve(tokensToRetrieve)
+		t.instanceTokenBucket.ForceRetrieve(tokensToRetrieve)
+		return nil
+	}
+
+	// if we can't retrieve from request bucket, check shared buckets
+	retrieved = t.userTokenBucket.Retrieve(tokensToRetrieve)
+	if !retrieved {
+		// if dry run, force retrieve all tokens and return nil
+		if t.dryRun {
+			t.requestTokenBucket.ForceRetrieve(tokensToRetrieve)
+			t.userTokenBucket.ForceRetrieve(tokensToRetrieve)
+			t.instanceTokenBucket.ForceRetrieve(tokensToRetrieve)
+			level.Warn(util_log.Logger).Log("msg", "not enough tokens in user token bucket", "dataType", dataType, "dataSize", num, "tokens", tokensToRetrieve)
+			return nil
+		}
+		return fmt.Errorf("not enough tokens in user token bucket")
+	}
+
+	retrieved = t.instanceTokenBucket.Retrieve(tokensToRetrieve)
+	if !retrieved {
+		t.userTokenBucket.Refund(tokensToRetrieve)
```

Comment:
We refund as the `Retrieve` function already deducts tokens from the bucket. Since we have series of buckets that we check, and if the last bucket fails, we need to revert back the tokens from other buckets

Reply:
It's okay if we don't return here too right?

Reply:
Ah I see, it would make the token bucket to limit more aggressively but I just wanted to take away same amount of tokens across all buckets (hence we don't deduct any tokens if one of the bucket check fails)